src/ast/binary_encoder.ml

```diff
@@ -0,0 +1,752 @@
+  Buffer.add_buffer buf exp_buf
+
+(* start: section 8 *)
+let encode_start buf int_opt =
+  match int_opt with None -> () | Some funcidx -> write_u32_of_int buf funcidx
+
+(* element: section 9 *)
+let encode_elements buf { Named.values = elems; _ } =
+  encode_vector buf elems (fun buf elem ->
+      let elem = Indexed.get elem in
+      write_element buf elem )
+
+(* datacount: section 12 *)
+let encode_datacount buf { Named.values = datas; _ } =
+  let len = List.length datas in
+  write_u32_of_int buf len
+
+(* code: section 10 *)
+let encode_codes buf funcs =
+  encode_vector buf funcs (fun buf { locals; body; _ } ->
+      let code_buf = Buffer.create 16 in
+      write_locals code_buf locals;
+      write_expr code_buf body ~end_op_code:None;
+      write_u32_of_int buf (Buffer.length code_buf);
+      Buffer.add_buffer buf code_buf )
+
+(* data: section 11 *)
+let encode_datas buf { Named.values = datas; _ } =
+  encode_vector buf datas (fun buf data ->
+      let data = Indexed.get data in
+      write_data buf data )
+
+let keep_local { Named.values; _ } =
+  List.filter_map
+    (fun data ->
+      match Indexed.get data with
+      | Runtime.Local data -> Some data
+      | Runtime.Imported _data -> None )
+    (List.rev values)
+
+let keep_imported { Named.values; _ } =
+  List.filter_map
+    (fun data ->
+      match Indexed.get data with
+      | Runtime.Local _data -> None
+      | Runtime.Imported data -> Some data )
+    (List.rev values)
+
+let encode (modul : Binary.modul) =
+  let buf = Buffer.create 256 in
+  let local_funcs = keep_local modul.func in
+  let local_tables = keep_local modul.table in
+  let local_memories = keep_local modul.mem in
+  let local_globales = keep_local modul.global in
+  let imported_funcs = keep_imported modul.func in
+  let imported_tables = keep_imported modul.table in
+  let imported_memories = keep_imported modul.mem in
+  let imported_globals = keep_imported modul.global in
+  Buffer.add_string buf "\x00\x61\x73\x6d";
+  (* magic *)
+  Buffer.add_string buf "\x01\x00\x00\x00";
+  (* version *)
+  encode_section buf '\x01' encode_types local_funcs;
+  encode_section buf '\x02' encode_imports
+    (imported_funcs, imported_tables, imported_memories, imported_globals);
+  encode_section buf '\x03' encode_functions local_funcs;
+  encode_section buf '\x04' encode_tables local_tables;
+  encode_section buf '\x05' encode_memories local_memories;
+  encode_section buf '\x06' encode_globals local_globales;
+  encode_section buf '\x07' encode_exports modul.exports;
+  encode_section buf '\x08' encode_start modul.start;
+  encode_section buf '\x09' encode_elements modul.elem;
+  encode_section buf '\x0C' encode_datacount modul.data;
+  encode_section buf '\x0A' encode_codes local_funcs;
+  encode_section buf '\x0B' encode_datas modul.data;
+  Buffer.contents buf
+
+let write_file filename content =
+  let filename = Fpath.filename filename in
+  let filename = filename ^ ".wasm" in
```

Comment:
This should not be `filename ^ ".wasm"`, otherwise it leads to filenames like `a.wat.wasm` whereas we want `a.wasm`.